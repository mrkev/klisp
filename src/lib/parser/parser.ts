// based on: https://github.com/jneen/parsimmon/blob/master/examples/lisp.js

import P from "parsimmon";

export type LangPos = {
  start: P.Index;
  end: P.Index;
};

export type LangType = {
  Expression: LangType["Symbol"] | LangType["Number"] | LangType["List"];
  Symbol: { kind: "symbol"; symbol: string; "@": LangPos };
  Number: { kind: "number"; number: number; "@": LangPos };
  List: { kind: "list"; list: Array<LangType["Expression"]>; "@": LangPos };
  Module: {
    kind: "module";
    exprs: Array<LangType["Expression"]>;
    "@": LangPos;
  };
};

const Lisp = P.createLanguage<LangType>({
  // An expression is just any of the other values we make in the language. Note
  // that because we're using `.createLanguage` here we can reference other
  // parsers off of the argument to our function. `r` is short for `rules` here.
  Expression: function (r) {
    return P.alt(r.Symbol, r.Number, r.List);
  },

  // IDENTIFIER
  // The basic parsers (usually the ones described via regexp) should have a
  // description for error message purposes.
  Symbol: function () {
    return P.regexp(/[a-zA-Z_\-\+\*\/][a-zA-Z0-9_-]*/)
      .desc("symbol")
      .mark()
      .map(({ start, end, value }) => ({
        kind: "symbol",
        symbol: value,
        "@": { start, end },
      }));
  },

  // Note that Number("10") === 10, Number("9") === 9, etc in JavaScript.
  // This is not a recursive parser. Number(x) is similar to parseInt(x, 10).
  Number: function () {
    return P.regexp(/[0-9]+/)
      .map(Number)
      .desc("number")
      .mark()
      .map(({ start, end, value }) => ({
        kind: "number",
        number: value,
        "@": { start, end },
      }));
  },

  // `.trim(P.optWhitespace)` removes whitespace from both sides, then `.many()`
  // repeats the expression zero or more times. Finally, `.wrap(...)` removes
  // the '(' and ')' from both sides of the list.
  List: function (r) {
    return r.Expression.trim(P.optWhitespace)
      .many()
      .wrap(P.string("("), P.string(")"))
      .mark()
      .map(({ start, end, value }) => ({
        kind: "list",
        list: value,
        "@": { start, end },
      }));
  },

  // A file in Lisp is generally just zero or more expressions.
  Module: function (r) {
    return r.Expression.trim(P.optWhitespace)
      .many()
      .mark()
      .map(({ start, end, value }) => ({
        kind: "module",
        exprs: value,
        "@": { start, end },
      }));
  },
});

export function tryParse(str: string) {
  return Lisp.Module.tryParse(str);
}
