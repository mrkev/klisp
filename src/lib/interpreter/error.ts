import Parsimmon from "parsimmon";
import { LangPos } from "../parser/parser";

export type Meta = Readonly<{
  start: Parsimmon.Index;
  end: Parsimmon.Index;
}>;

export class ScriptError extends Error {
  constructor(message: string) {
    super(message);
  }
}

export class ScriptPosError extends Error {
  readonly pos: Meta;
  constructor(message: string, pos: LangPos) {
    super(message);
    this.pos = pos;
  }
}
