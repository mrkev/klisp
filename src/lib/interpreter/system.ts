import { ScriptError } from "./error";

/**
 * System handles I/O, other APIs
 */
export class InterpreterSystem {
  readonly console: Console = new Console();
}

class Console {
  _log: (string | Error)[] = [];
  _fatalError: Error | ScriptError | null = null;

  clear() {
    this._log = [];
  }

  log(msgs: string[]) {
    console.log(...msgs);
    this._log.push(...msgs);
  }

  fail(err: Error) {
    this._fatalError = err;
  }

  getOut(): string {
    return this._log.join("\n");
  }
}
