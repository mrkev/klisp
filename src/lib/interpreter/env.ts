import { LangType } from "../parser/parser";
import { nullthrows } from "../utils";
import { ScriptError, ScriptPosError } from "./error";
import { ValType, v } from "./value";

// TODO: change ERRORS for more specific ones

/**
 * Context handles variable/ scoping, stack frames (TODO), etc
 */
export class Environment {
  static standard() {
    const scope = new ScopeStack();
    const env = new Environment(scope);
    env.scope.set("print", v.op("print"));
    env.scope.set("let", v.op("let"));
    env.scope.set("fun", v.op("fun"));
    env.scope.set("list", v.op("list"));
    env.scope.set("+", v.op("+"));
    env.scope.set("-", v.op("-"));
    env.scope.set("*", v.op("*"));
    env.scope.set("/", v.op("/"));
    return env;
  }

  static fromScope(scope: ScopeStack) {
    return new Environment(scope);
  }

  private constructor(readonly scope: ScopeStack) {}
}

export class ScopeStack {
  // we start with top-level map

  constructor(
    private readonly stack: Array<Map<string, ValType["Value"]>> = [new Map()]
  ) {}

  push() {
    this.stack.push(new Map());
  }

  pop() {
    if (this.stack.length === 1) {
      throw new Error("Can't pop last scope!");
    } else if (this.stack.length < 0) {
      throw new Error(`Invalid state, ${this.stack.length} scopes.`);
    } else {
      this.stack.pop();
    }
  }

  // to create closures
  clone() {
    const stack: Array<Map<string, ValType["Value"]>> = [];
    console.log(this.stack);
    for (const map of this.stack) {
      stack.push(new Map(map));
      console.log("pushing");
    }
    return new ScopeStack(stack);
  }

  private top() {
    return nullthrows(
      this.stack.at(-1),
      "impossible: always at least one scope in stack"
    );
  }

  // throws
  get(sym: LangType["Symbol"]): ValType["Value"] {
    for (let i = this.stack.length - 1; i > -1; i--) {
      const current = this.stack[i];
      if (!current.has(sym.symbol)) {
        continue;
      }
      return nullthrows(
        current.get(sym.symbol),
        "impossible: checked for existance above"
      );
    }

    throw new ScriptPosError(`Unbound variable: ${sym.symbol}`, sym["@"]);
  }

  // let is:
  // - constant, can't redefine
  // - can however overwrite things in higher scopes
  let(sym: LangType["Symbol"], val: ValType["Value"]) {
    const top = this.top();
    if (top.has(sym.symbol)) {
      throw new ScriptPosError("Cant overwrite constant: " + sym, sym["@"]);
    }

    top.set(sym.symbol, val);
  }

  // set is:
  // - mutable, can be overwritten
  set(key: string, value: ValType["Value"]) {
    this.top().set(key, value);
  }
}
