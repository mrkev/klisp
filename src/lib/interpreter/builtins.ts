import { exhaustive } from "../utils";
import { Environment } from "./env";
import { InterpreterSystem } from "./system";
import { ValType, v } from "./value";
import { evaluate, stringOfValue } from "./interpreter";
import { ScriptError, ScriptPosError } from "./error";
import { LangType } from "../parser/parser";

export type BuiltInOp =
  | "print"
  | "let"
  | "fun"
  | "list"
  | "+"
  | "-"
  | "*"
  | "/";

export function evaluateOp(
  op: BuiltInOp,
  exprs: Array<LangType["Expression"]>,
  env: Environment,
  system: InterpreterSystem
): ValType["Value"] {
  switch (op) {
    case "let": {
      // const [sym, val] = args;
      // env.scope.set()
      // TODO
      return v.nil();
    }

    // (lambda (alpha) (* 5 alpha))
    case "fun": {
      // need arg list and body
      if (exprs.length < 2) {
        console.log(exprs);
        throw new ScriptError("Ill-formed special form");
      }

      const [arglist, ...body] = exprs;
      if (arglist.kind !== "list") {
      }

      const symbs = argumentList(arglist);

      return v.closure(
        symbs.map((s) => s.symbol),
        body,
        env
      );
    }

    case "list": {
      const args = exprs.map((x) => evaluate(x, env, system));
      return v.list(args);
    }

    case "+": {
      const args = exprs.map((x) => evaluate(x, env, system));
      const nums = expectNumbers(args);
      const sum = nums.reduce((a, x) => a + x.number, 0);
      return v.number(sum);
    }

    case "*": {
      const args = exprs.map((x) => evaluate(x, env, system));
      const nums = expectNumbers(args);
      const prod = nums.reduce((a, x) => a * x.number, 1);
      return v.number(prod);
    }

    case "-": {
      const args = exprs.map((x) => evaluate(x, env, system));
      const nums = expectNumbers(args);
      if (nums.length == 0) {
        throw new ScriptError("Wrong number of arguments passed to procedure");
      } else if (nums.length == 1) {
        return v.number(-nums[0].number);
      } else {
        // map to number so we don't need to fiddle with the first acc
        const diff = nums.map((x) => x.number).reduce((a, x) => a - x);
        return v.number(diff);
      }
    }

    case "/": {
      const args = exprs.map((x) => evaluate(x, env, system));
      const nums = expectNumbers(args);
      if (nums.length == 0) {
        throw new ScriptError("Wrong number of arguments passed to procedure");
      } else if (nums.length == 1) {
        return v.number(1 / nums[0].number);
      } else {
        // map to number so we don't need to fiddle with the first acc
        const quot = nums.map((x) => x.number).reduce((a, x) => a / x);
        return v.number(quot);
      }
    }

    case "print": {
      const args = exprs.map((x) => evaluate(x, env, system));
      const strs = args.map((val) => stringOfValue(val));
      system.console.log(strs);
      return v.nil();
    }

    default:
      throw exhaustive(op);
  }
}

function expectNumbers(
  vals: Array<ValType["Value"]>
): Array<ValType["Number"]> {
  for (const val of vals) {
    if (val.kind !== "number") {
      throw new ScriptError("Expected 'number'");
    }
  }
  return vals as Array<ValType["Number"]>;
}

function expectSymbolNodes(
  vals: Array<LangType["Expression"]>
): Array<LangType["Symbol"]> {
  for (const val of vals) {
    if (val.kind !== "symbol") {
      throw new ScriptPosError("Expected symbol", val["@"]);
    }
  }
  return vals as Array<LangType["Symbol"]>;
}

function argumentList(expr: LangType["Expression"]): Array<LangType["Symbol"]> {
  switch (expr.kind) {
    case "number":
      throw new ScriptPosError(
        "Identifier or identifier list expected",
        expr["@"]
      );
    case "list":
      return expectSymbolNodes(expr.list);
    case "symbol":
      return [expr];
    default:
      throw exhaustive(expr);
  }
}
