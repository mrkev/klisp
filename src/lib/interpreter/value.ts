import { LangType } from "../parser/parser";
import { BuiltInOp } from "./builtins";
import { Environment, ScopeStack } from "./env";

export type ValType = {
  Value:
    | ValType["Number"]
    | ValType["Nil"]
    | ValType["List"]
    | ValType["Closure"]
    | ValType["BuiltInOp"];
  // data
  Number: { kind: "number"; number: number };
  Nil: { kind: "nil" };
  List: { kind: "list"; list: Array<ValType["Value"]> };
  // callables
  Closure: { kind: "closure"; closure: Closure };
  BuiltInOp: { kind: "op"; op: BuiltInOp };
};

export const v = {
  number(number: number): ValType["Number"] {
    return { kind: "number", number };
  },
  list(list: ValType["Value"][]): ValType["List"] {
    return { kind: "list", list };
  },
  nil(): ValType["Nil"] {
    return { kind: "nil" };
  },
  closure(
    lambda: Array<LangType["Expression"]>,
    env: Environment
  ): ValType["Closure"] {
    return { kind: "closure", closure: new Closure(lambda, env.scope.clone()) };
  },
  op(op: BuiltInOp): ValType["BuiltInOp"] {
    return { kind: "op", op };
  },
};

class Closure {
  constructor(
    readonly lambda: Array<LangType["Expression"]>,
    readonly scope: ScopeStack
  ) {}
}
