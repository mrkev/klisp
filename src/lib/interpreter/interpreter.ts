import { LangType } from "../parser/parser";
import { exhaustive } from "../utils";
import { evaluateOp } from "./builtins";
import { Environment } from "./env";
import { ScriptError, ScriptPosError } from "./error";
import { InterpreterSystem } from "./system";
import { ValType, v } from "./value";

export function interpret(
  module: LangType["Module"],
  context: Environment,
  system: InterpreterSystem
): ValType["Value"] | ScriptError | ScriptPosError | Error {
  try {
    let result: ValType["Value"] = v.nil();
    for (const expr of module.exprs) {
      result = evaluate(expr, context, system);
    }
    return result;
  } catch (e: unknown) {
    if (e instanceof ScriptError) {
      return e;
    } else if (e instanceof Error) {
      return e;
    } else {
      return new Error(String(e));
    }
  }
}

export function evaluate(
  expr: LangType["Expression"],
  env: Environment,
  system: InterpreterSystem
): ValType["Value"] {
  switch (expr.kind) {
    case "number": {
      return v.number(expr.number);
    }

    case "list": {
      return evaluateList(expr, env, system);
    }

    case "symbol":
      return env.scope.get(expr);

    default:
      throw exhaustive(expr);
  }
}

// (value cdr cdr cdr...)
function evaluateList(
  expr: LangType["List"],
  env: Environment,
  system: InterpreterSystem
) {
  if (expr.list.length < 1) {
    throw new ScriptPosError("Ill-formed expression", expr["@"]);
  }

  const [car, ...cdr] = expr.list;
  const first = evaluate(car, env, system);
  switch (first.kind) {
    case "list":
    case "nil":
    case "number":
      throw new ScriptPosError("Operator is not a PROCEDURE", expr["@"]);

    //
    case "closure": {
      const envdown = Environment.fromScope(first.closure.scope);
      envdown.scope.push();
      if (cdr.length !== first.closure.args.length) {
        throw new ScriptError("Wrong number of arguments passed to procedure");
      }

      for (let i = 0; i < cdr.length; i++) {
        const argsym = first.closure.args[i];
        const argval = evaluate(cdr[i], env, system);
        envdown.scope.set(argsym, argval);
      }

      // note, lambdas always have at leastone expression
      let result: ValType["Value"] = v.nil();
      for (const expr of first.closure.body) {
        result = evaluate(expr, envdown, system);
      }

      // technically not necessary, but doing it for the sake fo symmetry
      envdown.scope.pop();
      return result;
    }

    case "op": {
      console.log(car, cdr);
      return evaluateOp(first.op, cdr, env, system);
    }
    default:
      throw exhaustive(first);
  }
}

export function stringOfValue(val: ValType["Value"]): string {
  switch (val.kind) {
    case "closure":
      return `(fun (${val.closure.args.join(" ")}) ...)`;
    case "list":
      return `(${val.list.map((v) => stringOfValue(v)).join(" ")})`;
    case "nil":
      return "nil";
    case "number":
      return String(val.number);
    case "op":
      return `<builtin op: ${val.op}>`;

    default:
      throw exhaustive(val);
  }
}
