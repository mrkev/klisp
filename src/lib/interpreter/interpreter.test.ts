import { describe, expect, it } from "vitest";
import { tryParse } from "../parser/parser";
import { interpret, stringOfValue } from "./interpreter";
import { Environment } from "./env";
import { InterpreterSystem } from "./system";
import { ValType } from "./value";

function run(str: string): ValType["Value"] {
  const ast = tryParse(str);

  const context = Environment.standard();
  const system = new InterpreterSystem();
  const result = interpret(ast, context, system);

  if (result instanceof Error) {
    throw result;
  }

  return result;
}

function testOk(testname: string, script: string, res: string) {
  it(testname, () => {
    const value = run(script);
    expect(stringOfValue(value)).toBe(res);
  });
}

function testErr(testname: string, script: string) {
  it(testname, () => {
    expect(() => run(script)).toThrowErrorMatchingSnapshot();
  });
}

describe("math", () => {
  testOk("+", "(+ 3 4)", "7");
  testOk("-", "(- 3 4)", "-1");
  testOk("- unary", "(- 3)", "-3");
  testOk("*", "(* 3 4)", "12");
  testOk("/", "(/ 4 2)", "2");
  testOk("/ unary", "(/ 4)", "0.25");
});

describe("list", () => {
  testOk("list", "(list 3 4)", "(3 4)");
});

describe("lambda", () => {
  testOk("id", "(fun x x)", "(fun (x) ...)");
  testOk("id.2", "(fun (x) x)", "(fun (x) ...)");
  testOk("apply", "((fun (x) x) 3)", "3");
  testErr("x.args.1", "((fun (x) x) 3 3)");
  testErr("x.args.2", "((fun (x y) (+ x y)) 3)");
  testOk("body list", "((fun (x y) (+ x y) (- x y)) 3 2)", "1");
});
