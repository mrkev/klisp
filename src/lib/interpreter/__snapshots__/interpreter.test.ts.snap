// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lambda > x.args.1 1`] = `[Error: Wrong number of arguments passed to procedure]`;

exports[`lambda > x.args.2 1`] = `[Error: Wrong number of arguments passed to procedure]`;
